### ISSUE TITLE
Add LapRF Interface

### ISSUE THREAD

**Issue by johnhooks** [RotorHazard/RotorHazard#560]
I am working on adding a LapRF Interface to RotorHazard, though after looking through `src/interface/RHInterface.py` it looks like it could take some modification to make the two fit together, they work rather differently. 

#### Some of the differences

- The LapRF calculates when to trigger passing events internally.
- The LapRF uses two settings per receiver`threshold` and `gain` to determine when to trigger a passing event.
- The LapRF sends status updates with `last_rssi` per receiver. I believe the interval at which these are sent could be adjusted (the default interval is 1 second), though I haven't tried it yet and it most likely still not provide enough information to edit the race results the same way as the RotorHazard timer.

Would it make more sense to unify the api of interfaces by creating something more generic than `BaseHardwareInterface`? Or to run two separate interfaces side by side?

My first impression is that the `RHInterface` is tightly coupled to how the current system runs and to integrate the LapRF will be difficult without making it a total separate system with its own initialization, UI, and integration with `server.py`.

#### A quick explanation of how the LapRF works

It sends and receives packets that contain one or more records. The records contain a header with information about the record and multiple fields containing the record's data.

The types of records: rssi, rfSetup, stateControl, settings, descriptor, passing, status, time, error.

It automatically sends **status** and **passing** records without a request.
 
A top level LapRFInterface would manage connection to multiple LapRF devices, of either serial or network connection. A LapRFDevice would look something like this.

```python
class LapRFDevice():
    def milliseconds():
        """"An adjusted time based off the RTC on the LapRF device
        """"
        pass
    def get_node_rfsetup(index):
        pass
    def set_node_rfsetup(index, enabled, channel, band, threshold, gain):
        pass
    def get_min_lap_time():
        pass
    def set_min_lap_time(milliseconds):
        pass
    def update():
        """Read status and passing records from the LapRF and relay the information
        """
        pass
```

Any suggestions on how best to accomplish the goal would be greatly appreciated.

**Comment by HazardCreative** [comment]
RotorHazard should not set a minimum lap time on the LapRF (or any other hardware). Using min lap will prevent possible valid passes from reaching the server (such as missing the gate and looping back around to pass through it). RotorHazard highlights short laps for the operator to review. For those who wish to have the min lap functionality, RH can already automatically discard them server-side. Doing this on the server means the passing data still exists and can be recovered if needed (vs. when implemented on hardware the data is never sent and lost forever.)

**Comment by johnhooks** [comment]
@HazardCreative I will try to set the minimum lap time to 0 and remove it from the interface. I'll need to experiment and make sure it doesn't do something to the LapRF device, I don't have access to how it handles passing events internally. If it can't be set to zero I will set it as low as possible, and make notes about it in the code.

#### Updated LapRFDevice 

```python
class LapRFDevice():
    def milliseconds(): # An adjusted time based off the RTC on the LapRF device
        pass
    def get_node_rfsetup(index):
        pass
    def set_node_rfsetup(index, enabled, channel, band, threshold, gain):
        pass
    def update(): # Read status and passing records from the LapRF and relay the information
        pass
```

**Comment by pulquero** [comment]
A while back, I did a similar exercise for Chorus (which is quite different from RH). These two commits should help: 
https://github.com/pulquero/RotorHazard/commit/f3461ec23102c8c824e82f64f40e8c3fbafd0096
https://github.com/pulquero/RotorHazard/commit/e7ec6d8ba2e6aebced98aaf8f51b57c750528488

**Comment by ethomas997** [comment]
One thing I'd recommend is that if you're looking to change how the core system operates, try to break up changes into multiple pieces (with separate pull requests), rather than try to develop and post a major rework all at once.  Smaller changes are much easier to deal with and are more likely to be accepted.

**Comment by johnhooks** [comment]
@ethomas997 thanks for the advice, right now I am working on the basic LapRFDevice and the record encode/decode functions. After those are done I'll post back in here with a reference to my fork and you guys can take a look and let me know what you think about the basics.

**Comment by johnhooks** [comment]
There are two different ways to connect a LapRF timer, over a network socket or serial. I've been familiarizing myself with `gevent`, I think I understand what it does with sockets, the calls to IO functions yield to the rest of the program until able to read or write. Though for a serial connection, I see that `gevent.lock.RLock` is used in `src/interface/serial_node.py`. Is this because `pyserial` uses the `select` module and that is modified by the `gevent.money.patch_all()`? Without the Rlock it would yield and very possibly a different greenlet would attempt a access the serial connection in the middle of another operation?

**Comment by ethomas997** [comment]
@johnhooks Yes -- for the serial connection to the S32_BPill board with multiple nodes, the currently-addressed node is set via 'check_set_multi_node_index()' and then the read or write is performed, so it's important that another thread does not jump in during the operation.

**Comment by johnhooks** [comment]
I have almost complete the LapRF protocol. I have pushed it into the branch `laprf-interface` in my fork of RotorHazard. These are the added files. I have type hints in there, if you guys don't want them I can remove them, but it does improve the IDE experience for me, with much better autocomplete for instance property names. These are the list of relevant files:

- [src/interface/laprf-protocol.py](https://github.com/johnhooks/RotorHazard/blob/37a56457c7d01890db279d68ab471af4898b582d/src/interface/laprf_protocol.py)
- [src/tests/laprf-mock.py](https://github.com/johnhooks/RotorHazard/blob/37a56457c7d01890db279d68ab471af4898b582d/src/tests/laprf_mock.py)
- [src/tests/laprf-protocol.py](https://github.com/johnhooks/RotorHazard/blob/37a56457c7d01890db279d68ab471af4898b582d/src/tests/test_laprf_protocol.py)

Next, I will begin work on how to connect to both socket and serial devices.

Also, I have never performed a pull request before, so I will need to be walked through it when ever that time comes.

**Comment by ethomas997** [comment]
For GitHub PR info please see:  https://github.com/RotorHazard/RotorHazard/blob/main/doc/Development.md

**Comment by arvtom** [comment]
@johnhooks what's the status on this issue at the moment ? I would be glad to help with lapRF protocol. I have 8-way laprf and RPI4 available for testing.

**Comment by johnhooks** [comment]
@arvtom I have a LapRF encoder and decoder that is written and tested, I started diving in to how to integrate it with RotorHazard and hit a wall. I tried to familiarize myself with how RotorHazard nodes communicate the data to the main application but I had trouble understanding how the LapRF would connect in.

**Comment by martinglass** [comment]
In case you've not seen this already, this fork of RH has the LapRF option built in.
https://github.com/pulquero/RotorHazard

Read the setup guide,because the install requirements are different to the original RH code, including https://github.com/pulquero/RotorHazard#quick-start 

Note: LapRF IP address goes into config.json.
